add dbt_valid_to_current (#6308)
fishtown-analytics/docs.getdbt.com: website/docs/reference/resource-configs/dbt_valid_to_current.md

---
resource_types: [snapshots]
description: "Snapshot dbt_valid_to_current custom date"
datatype: "{<dictionary>}"
default_value: {NULL}
id: "dbt_valid_to_current"
---

Available in 1.9 or with [Versionless](/docs/dbt-versions/upgrade-dbt-version-in-cloud#versionless) dbt Cloud.

<File name='snapshots/schema.yml'>

```yaml
snapshots:
  my_project:
    +dbt_valid_to_current: "to_date('9999-12-31')"

```

</File>

<File name='snapshots/<filename>.sql'>

```sql
{{
    config(
        unique_key='id',
        strategy='timestamp',
        updated_at='updated_at',
        dbt_valid_to_current='to_date('9999-12-31')'
    )
}}
```

</File>

<File name='dbt_project.yml'>

```yml
snapshots:
  [<resource-path>](/reference/resource-configs/resource-path):
    +dbt_valid_to_current: "to_date('9999-12-31')"
```

</File>

## Description

Use the `dbt_valid_to_current` config to set a custom future date for `dbt_valid_to` in new snapshot columns. When set, dbt will use this specified value instead of `NULL` for `dbt_valid_to` in the snapshot table.

This approach makes it easier to assign a custom date date, work in a join, or perform range-based filtering that require an end date.

> ## Usage

Author: ```suggestion
### Considerations
```

---
resource_types: [snapshots]
description: "Snapshot dbt_valid_to_current custom date"
datatype: "{<dictionary>}"
default_value: {NULL}
id: "dbt_valid_to_current"
---

> 

Reviewer: I don't know where we want this, but we should probably have a [migration callout/warning](https://docs.getdbt.com/reference/resource-configs/snapshot_meta_column_names#:~:text=To%20avoid%20any,column%20name%20change.) like what we have for the other new configs cc: @dbeatty10

Available in 1.9 or with [Versionless](/docs/dbt-versions/upgrade-dbt-version-in-cloud#versionless) dbt Cloud.

<File name='snapshots/schema.yml'>

```yaml
snapshots:
  my_project:
    +dbt_valid_to_current: "to_date('9999-12-31')"

```

</File>

<File name='snapshots/<filename>.sql'>

```sql
{{
    config(
        unique_key='id',
        strategy='timestamp',
        updated_at='updated_at',
        dbt_valid_to_current='to_date('9999-12-31')'
    )
}}
```

</File>

<File name='dbt_project.yml'>

```yml
snapshots:
  [<resource-path>](/reference/resource-configs/resource-path):
    +dbt_valid_to_current: "to_date('9999-12-31')"
```

</File>

## Description

> Use the `dbt_valid_to_current` config to set a custom future date for `dbt_valid_to` in new snapshot columns. When set, dbt will use this specified value instead of `NULL` for `dbt_valid_to` in the snapshot table.

Reviewer: again, rephrase as suggested above

thinking about this more, i **do** like that we're calling out that the main use case for this is "a future date"

> This approach makes it easier to assign a custom date date, work in a join, or perform range-based filtering that require an end date.

Reviewer: "date date"

> This approach makes it easier to assign a custom date date, work in a join, or perform range-based filtering that require an end date.

Author: ```suggestion
This approach makes it easier to assign a custom date, work in a join, or perform range-based filtering that requires an end date.
```

- **Jinja limitation** &mdash; `dbt_valid_to_current` only accepts static SQL expressions. Jinja expressions (like `{{ var('my_future_date') }}`) are not supported.

- **Deferral and `state:modified`** &mdash; Changes to `dbt_valid_to_current` are compatible with deferral and `--select state:modified`. When this configuration changes, it'll appear in `state:modified` selections, raising a warning to manually make the necessary snapshot updates.

## Default

By default, `dbt_valid_to` is set to `NULL` for current (most recent) records in your snapshot table. This means that these records are still valid and have no defined end date.

If you prefer to use a specific value instead of `NULL` for `dbt_valid_to` in current and future records, you can use the `dbt_valid_to_current` configuration option. For example, setting a date in the far future, `9999-12-31`.

The value assigned to `dbt_valid_to_current` should be a string representing a valid date or timestamp, depending on your database's requirements. Use expressions that work within the data platform.

> ### Managing records

Reviewer: Is there a more specific descriptor here for the heading?

> - **For existing records** &mdash; To avoid any unintentional data modification, dbt will _not_ automatically adjust the current value in the existing `dbt_valid_to` column. Existing current records will still have `dbt_valid_to` set to `NULL`.

Reviewer: Ah here's the migration callout - I think we just want to make sure this warning exists elsewhere as well!

- **For new records** &mdash;  Any new records inserted after applying the `dbt_valid_to_current` configuration will have `dbt_valid_to` set to the specified value (for example, '9999-12-31'), instead of `NULL`.

> This means your snapshot table will have current records with `dbt_valid_to` values of both `NULL` (from existing data) and the new specified value (from new data). If you'd rather have consistent `dbt_valid_to` values for current records, you can either manually update existing records in your snapshot table where `dbt_valid_to` is `NULL` to match your `dbt_valid_to_current` value or rebuild your snapshot table.

Reviewer: "rebuild your snapshot table" would be very risky (lose all of your historical data!)

---
resource_types: [snapshots]

> description: "Snapshot dbt_valid_to_current custom date"

Reviewer: should we update this to custom current indicator?

datatype: "{<dictionary>}"
default_value: {NULL}
id: "dbt_valid_to_current"
---

> Available in 1.9 or with [Versionless](/docs/dbt-versions/upgrade-dbt-version-in-cloud#versionless) dbt Cloud.

Author: ```suggestion
Available from dbt v1.9 or with [Versionless](/docs/dbt-versions/upgrade-dbt-version-in-cloud#versionless) dbt Cloud.
```

<File name='snapshots/schema.yml'>

```yaml
snapshots:
  my_project:
    +dbt_valid_to_current: "to_date('9999-12-31')"

```

</File>

<File name='snapshots/<filename>.sql'>

```sql
{{
    config(
        unique_key='id',
        strategy='timestamp',
        updated_at='updated_at',
        dbt_valid_to_current='to_date('9999-12-31')'
    )
}}
```

</File>

<File name='dbt_project.yml'>

```yml
snapshots:
  [<resource-path>](/reference/resource-configs/resource-path):
    +dbt_valid_to_current: "to_date('9999-12-31')"
```

</File>

## Description

> Use the `dbt_valid_to_current` config to set a custom indicator for the value of `dbt_valid_to` in current snapshot records (like a future date). By default, this value is `NULL`. When set, dbt will use this specified value instead of `NULL` for `dbt_valid_to` in the snapshot table.

Reviewer: ```suggestion
Use the `dbt_valid_to_current` config to set a custom indicator for the value of `dbt_valid_to` in current snapshot records (like a future date). By default, this value is `NULL`. When set, dbt will use this specified value instead of `NULL` for `dbt_valid_to` for current records in the snapshot table.
```

If you prefer to use a specific value instead of `NULL` for `dbt_valid_to` in current and future records, you can use the `dbt_valid_to_current` configuration option. For example, setting a date in the far future, `9999-12-31`.

The value assigned to `dbt_valid_to_current` should be a string representing a valid date or timestamp, depending on your database's requirements. Use expressions that work within the data platform.

## Impact on snapshot records

When you set `dbt_valid_to_current`, it affects how dbt manages the `dbt_valid_to` column in your snapshot table:

- **For existing records** &mdash; To avoid any unintentional data modification, dbt will _not_ automatically adjust the current value in the existing `dbt_valid_to` column. Existing current records will still have `dbt_valid_to` set to `NULL`.

- **For new records** &mdash;  Any new records inserted after applying the `dbt_valid_to_current` configuration will have `dbt_valid_to` set to the specified value (for example, '9999-12-31'), instead of `NULL`.

> This means your snapshot table will have current records with `dbt_valid_to` values of both `NULL` (from existing data) and the new specified value (from new data). If you'd rather have consistent `dbt_valid_to` values for current records, you can manually update existing records in your snapshot table where `dbt_valid_to` is `NULL` to match your `dbt_valid_to_current` value.

Author: ```suggestion
This means your snapshot table will have current records with `dbt_valid_to` values of both `NULL` (from existing data) and the new specified value (from new data). If you'd rather have consistent `dbt_valid_to` values for current records, you can manually update existing records in your snapshot table (where `dbt_valid_to` is `NULL`) to match your `dbt_valid_to_current` value.
```

---
resource_types: [snapshots]
description: "Snapshot dbt_valid_to_current custom date"
datatype: "{<dictionary>}"

> default_value: {"dbt_valid_from": "dbt_valid_from", "dbt_valid_to": "dbt_valid_to", "dbt_scd_id": "dbt_scd_id", "dbt_updated_at": "dbt_updated_at"}

Reviewer 2: The default value for `dbt_valid_to_current` is `NULL`.

I'm guessing this was a copy-paste carry-over from the docs for `snapshot_meta_column_names`.

Author: good shout!

id: "dbt_valid_to_current"
---

Available in 1.9 or with [Versionless](/docs/dbt-versions/upgrade-dbt-version-in-cloud#versionless) dbt Cloud.

<File name='snapshots/schema.yml'>

```yaml
snapshots:
  my_project:
    +dbt_valid_to_current: "to_date('9999-12-31')"

```

</File>

<File name='snapshots/<filename>.sql'>

```sql
{{
    config(
        unique_key='id',
        strategy='timestamp',
        updated_at='updated_at',
        dbt_valid_to_current='to_date('9999-12-31')'
    )
}}
```

</File>

<File name='dbt_project.yml'>

```yml
snapshots:
  [<resource-path>](/reference/resource-configs/resource-path):
    +dbt_valid_to_current: "to_date('9999-12-31')"
```

</File>

## Description

Use the `dbt_valid_to_current` config to set a custom future date for `dbt_valid_to` in new snapshot columns. When set, dbt will use this specified value instead of `NULL` for `dbt_valid_to` in the snapshot table.

> This simplifies makes it easier to assign date, work in a join, and range-based filtering with an end date.

Reviewer 2: This sentence sounds funny, but I don't know how to re-work it. But I bet you'll have some ideas 🧠

## Default

By default, `dbt_valid_to` is set to `NULL` for current (most recent) records in your snapshot table. This means that these records are still valid and have no defined end date.

If you prefer to use a specific value instead of `NULL` for `dbt_valid_to` in current and future records, you can use the `dbt_valid_to_current` configuration option. For example, setting a date in the far future, `9999-12-31`.

> The value assigned to `dbt_valid_to_current` should be a string representing a valid date or timestamp, depending on your database's requirements.

Author: ```suggestion
The value assigned to `dbt_valid_to_current` should be a string representing a valid date or timestamp, depending on your database's requirements. Use expressions that work within the data platform.
```
